### ISSUE TITLE
substraction overflow in NewMmapLayout

### ISSUE THREAD

**Issue by CharlyYu** [QuarkContainer/Quark#1019]
```
[Print] [1/43c0180000|0] get panic : Some(attempt to subtract with overflow)
[INFO] [1/493379] hypecall: 8, paras: 43c019a128 0 0 0
[Print] [1/43c0180000|0] panic occurred in file 'src/qlib/kernel/arch/./aarch64/context.rs' at line 137
```

when compared with x86_64 codes, it seems the difference lies in this constant definition

```
const MMAP_RND_BITS: u64 = 33;
pub const MAX_MMAP_RAND64: u64 = (1u64 << MMAP_RND_BITS) << MemoryDef::PAGE_SHIFT;
```

In x86_64, it is
```
pub const MAX_MMAP_RAND64: u64 = (1 << 28) * MemoryDef::PAGE_SIZE;
```

I don't quite get the background, Could you please take a look at this?  @chl337

**Comment by chl337** [comment]
@CharlyYu is your kernel built with _CONFIG_ARM64_VA_BITS_48_?
Could you post the results of any of the below points:

1. if in the Linux .config file the _CONFIG_IKCONFIG_ is set, result of `zgrep -E "CONFIG_ARM64_VA_BITS_" /proc/config.gz`
2. else in the Linux .config file result of `grep "CONFIG_PGTABLE_LEVELS" .config ` and in the terminal result of `getconf PAGESIZE`.

**Comment by CharlyYu** [comment]
```
root@raspberrypi:/home/feng# zgrep -E "CONFIG_ARM64_VA_BITS_" /proc/config.gz
gzip: /proc/config.gz: No such file or directory
root@raspberrypi:/boot# uname -a
Linux raspberrypi 5.15.0-1041-raspi #44-Ubuntu SMP PREEMPT Tue Oct 10 13:05:59 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
root@raspberrypi:/boot# grep "CONFIG_ARM64_VA_BITS_" config-5.15.0-1041-raspi
# CONFIG_ARM64_VA_BITS_39 is not set
CONFIG_ARM64_VA_BITS_48=y
root@raspberrypi:/home/feng# getconf PAGESIZE
4096

```

**Comment by chl337** [comment]
> when compared with x86_64 codes, it seems the difference lies in this constant definition

The constants, both for x86_64 and aarch64, refer back to Linux.
Can you please try this [update](https://jmp.sh/GJQmVF4G).

**Comment by CharlyYu** [comment]
```
[DEBUG] [1/43c0180000|0] VM: NewMmapLayout - min addr 1000, max addr 140737488351232
[INFO] [1/438305] hypecall: 8, paras: 43c0199f48 0 0 0
[Print] [1/43c0180000|0] get panic : Some(attempt to subtract with overflow)
[INFO] [1/438402] hypecall: 8, paras: 43c0199f48 0 0 0
[Print] [1/43c0180000|0] panic occurred in file 'src/qlib/kernel/arch/./aarch64/context.rs' at line 139
```

still panic @chl337

**Comment by chl337** [comment]
I see the issue, it lays way back to memory size definitions. It seems as an inherited difference from Linux btw [x86_64](https://www.kernel.org/doc/Documentation/x86/x86_64/mm.txt) and [aarch64](https://www.kernel.org/doc/html/v5.8/arm64/memory.html). I would suggest to apply this [patch](https://jmp.sh/d42v4OvK) first. Please let me know of any changes.

**Comment by chl337** [comment]
@CharlyYu can you please give feedback on this?

**Comment by shrik3** [comment]
fixed in #1027
